sqlx-core/src/pool/inner.rs

```diff
@@ -409,3 +425,85 @@ impl<DB: Database> Drop for DecrementSizeGuard<DB> {
         self.pool.semaphore.release(1);
     }
 }
+
+use pin_project::{pin_project, pinned_drop};
+
+/// A [`PollWallClockRecorder`] measures the duration of time (in milliseconds)
+/// from first poll of `F`, to drop of this type.
+///
+/// The typical lifecycle of a future is `create -> poll -> ready -> drop`, with
+/// the drop happening an insignificant duration of time after the future
+/// returns [`Poll::Ready`].
+///
+/// Another common lifecycle path that should be instrumented is when the
+/// created future's caller times out before returning [`Poll::Ready`], with a
+/// lifecycle similar to `create -> poll -> pending -> timeout -> drop`. This
+/// results in the instrumented future being dropped an insignificant duration
+/// of time after the timeout occurs.
+///
+/// This instrumentation therefore (approximately) measures the time between
+/// first poll and yielding a value, or the future timing out and being dropped.
+#[pin_project(PinnedDrop)]
+struct PollWallClockRecorder<'a, F> {
+    // The atomic value that will have the poll duration (in milliseconds) added
+    // to it.
+    recorder: &'a AtomicU64,
+
+    // The instant of the first poll.
+    //
+    // None at construction time, Some after first poll.
+    started_at: Option<Instant>,
+
+    // The inner future that requires pinning to call poll().
+    #[pin]
+    fut: F,
+}
+
+impl<'a, F> PollWallClockRecorder<'a, F> {
+    /// Record the poll duration of `F`, adding the duration of time between
+    /// first poll and drop of `Self` in milliseconds, to `recorder`.
+    pub(crate) fn new(recorder: &'a AtomicU64, fut: F) -> Self {
+        Self {
+            recorder,
+            fut,
+            started_at: None,
+        }
+    }
+}
+
+impl<'a, F> Future for PollWallClockRecorder<'a, F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // Initialise the `started_at` instant on the first poll.
+        //
+        // This is more accurate than initialising the Instant in the
+        // constructor of this type, which would incorrectly include the time
+        // from construction to first poll in the measurement (which also
+        // includes the time a task waits to be scheduled onto a runtime worker
+        // thread).
+        let _started_at = this.started_at.get_or_insert_with(Instant::now);
+
+        // Pass through the poll to the inner future.
+        this.fut.poll(cx)
```

Comment:
We could inspect the result of the poll here and emit the elapsed duration when we see a `Poll::ready()` instead of at drop time for a more accurate measurement, but given the usage & negligible time difference I thought it was best to go for a simple impl first - though I'm happy to build it up.